cluster/mesos/cluster.go

```diff
@@ -0,0 +1,453 @@
+	return container
+}
+
+// Container returns the container with IdOrName in the cluster
+func (c *Cluster) Container(IDOrName string) *cluster.Container {
+	// Abort immediately if the name is empty.
+	if len(IDOrName) == 0 {
+		return nil
+	}
+
+	c.RLock()
+	defer c.RUnlock()
+
+	return formatContainer(cluster.Containers(c.Containers()).Get(IDOrName))
+}
+
+// RemoveImage removes an image from the cluster
+func (c *Cluster) RemoveImage(image *cluster.Image) ([]*dockerclient.ImageDelete, error) {
+	return nil, errNotSupported
+}
+
+// Pull will pull images on the cluster nodes
+func (c *Cluster) Pull(name string, authConfig *dockerclient.AuthConfig, callback func(what, status string)) {
+
+}
+
+// Load images
+func (c *Cluster) Load(imageReader io.Reader, callback func(what, status string)) {
+
+}
+
+// RenameContainer Rename a container
+func (c *Cluster) RenameContainer(container *cluster.Container, newName string) error {
+	//FIXME this doesn't work as the next refreshcontainer will erase this change (this change is in-memory only)
+	container.Config.Labels[cluster.SwarmLabelNamespace+".mesos.name"] = newName
+
+	return nil
+}
+
+func scalarResourceValue(offers map[string]*mesosproto.Offer, name string) float64 {
+	var value float64
+	for _, offer := range offers {
+		for _, resource := range offer.Resources {
+			if *resource.Name == name {
+				value += *resource.Scalar.Value
+			}
+		}
+	}
+	return value
+}
+
+// listNodes returns all the nodess in the cluster.
+func (c *Cluster) listNodes() []*node.Node {
+	c.RLock()
+	defer c.RUnlock()
+
+	out := []*node.Node{}
+	for _, s := range c.slaves {
+		n := node.NewNode(s.engine)
+		n.ID = s.id
+		n.TotalCpus = int64(scalarResourceValue(s.offers, "cpus"))
+		n.UsedCpus = 0
+		n.TotalMemory = int64(scalarResourceValue(s.offers, "mem")) * 1024 * 1024
+		n.UsedMemory = 0
+		out = append(out, n)
+	}
+	return out
+}
+
+func (c *Cluster) listOffers() []*mesosproto.Offer {
+	c.RLock()
+	defer c.RUnlock()
+
+	list := []*mesosproto.Offer{}
+	for _, s := range c.slaves {
+		for _, offer := range s.offers {
+			list = append(list, offer)
+		}
+	}
+	return list
+}
+
+// Info gives minimal information about containers and resources on the mesos cluster
+func (c *Cluster) Info() [][2]string {
+	offers := c.listOffers()
+	info := [][2]string{
+		{"\bStrategy", c.scheduler.Strategy()},
+		{"\bFilters", c.scheduler.Filters()},
+		{"\bOffers", fmt.Sprintf("%d", len(offers))},
+	}
+
+	sort.Sort(offerSorter(offers))
+
+	for _, offer := range offers {
+		info = append(info, [2]string{" Offer", offer.Id.GetValue()})
+		for _, resource := range offer.Resources {
+			info = append(info, [2]string{"  └ " + *resource.Name, fmt.Sprintf("%v", resource)})
+		}
+	}
+
+	return info
+}
+
+func (c *Cluster) addOffer(offer *mesosproto.Offer) {
+	s, ok := c.slaves[offer.SlaveId.GetValue()]
+	if !ok {
+		return
+	}
+	s.addOffer(offer)
+	go func(offer *mesosproto.Offer) {
+		<-time.After(c.offerTimeout)
+		if c.removeOffer(offer) {
```

Comment:
So, after a while we `removeOffer` the offer all the time?

What does it do exactly?